### PR TITLE
Better timeline

### PR DESCRIPTION
**Pull request by panyx0718**
1. Supported nested event
2. Better support for cpu multi-thread
3. Betters support nested block and block total time.

### CODE REVIEW

File: paddle/fluid/platform/profiler.cc [hunk 1]
```diff
@@ -147,19 +147,48 @@ RecordEvent::RecordEvent(const std::string& name, const DeviceContext* dev_ctx)
...
+
+RecordThread::RecordThread(int thread_id) {
+  if (g_state == ProfilerState::kDisabled) return;
+  SetCurThread(thread_id);
+}
+
+RecordThread::~RecordThread() {
+  if (g_state == ProfilerState::kDisabled) return;
+  ClearCurThread();
+}
```
Comment (chengduoZH) [inline paddle/fluid/platform/profiler.cc]: It seems that `RecordThread` only supports one layer nesting. If `RecordThread` records the `PreThread` before setting `CurThread`, `RecordThread` can support more layers nesting.

```  
   thread0
        thread1
             thread2
        thread1
   thread0
```

Comment (panyx0718) [reply paddle/fluid/platform/profiler.cc]: Yes, currently, it doesn't auto support thread nesting. See comments here:
paddle/fluid/platform/device_tracer.cc
The main reason is that we don't have nest threading yet.

Actually, user can add code to give nested thread different ids:

int nested_thread_id = CurThread() * 10 + 1
std::thread([nested_thread_id]() {
  RecordThread(nested_thread_id);
  ...
}

File: paddle/fluid/framework/executor.cc [hunk 2]
```diff
@@ -33,6 +34,11 @@ DEFINE_bool(check_nan_inf, false,
 
 namespace paddle {
 namespace framework {
+namespace {
+// block id starts from 0. This id is used to represent the codeblock
+// wrapping the first block 0.
+int kProgramId = -1;
```
Comment (chengduoZH) [inline paddle/fluid/framework/executor.cc]: Is `kProgramId` to distinguish the different `program` in a model?

Comment (panyx0718) [reply paddle/fluid/framework/executor.cc]: No. The global block has block id 0. However, before the block 0 starts, there are some other preprocessing and post-processing overhead. I wrap those codes into block id -1.

**Pull request merged**